### PR TITLE
identify and remove putative doublets

### PR DESCRIPTION
**Pull request by DavidBJaffe**
1. Add a doublet test, which removes some exact subclonotypes that appear to represent doublets.
2. Improve the disjoint orbit test.

General effect (see below): this change results in the removal of a significant number of three and
four chain clonotypes.  It makes `enclone` a bit slower.

```
test over 462 datasets (enclone.test)

BEFORE

   ┌────────┬────────────────────────┬──────────────────┬───────┐
   │chains  │  clonotypes with this  │  cells in these  │      %│
   │        │      number of chains  │      clonotypes  │       │
   ├────────┼────────────────────────┼──────────────────┼───────┤
   │1       │                 40914  │           44894  │    7.7│
   │2       │                432439  │          489439  │   84.0│
   │3       │                 24638  │           26914  │    4.6│
   │4       │                 20945  │           21341  │    3.7│
   │5       │                     5  │              92  │    0.0│
   │total   │                518941  │          582680  │  100.0│
   └────────┴────────────────────────┴──────────────────┴───────┘

AFTER

   ┌────────┬────────────────────────┬──────────────────┬───────┐
   │chains  │  clonotypes with this  │  cells in these  │      %│
   │        │      number of chains  │      clonotypes  │       │
   ├────────┼────────────────────────┼──────────────────┼───────┤
   │1       │                 40914  │           44894  │    7.7│
   │2       │                432442  │          489506  │   84.3│
   │3       │                 24018  │           26218  │    4.5│
   │4       │                 19877  │           20256  │    3.5│
   │5       │                     5  │              92  │    0.0│
   │total   │                517256  │          580966  │  100.0│
   └────────┴────────────────────────┴──────────────────┴───────┘
```

The change fixes a large many-chain clonotype in a customer dataset.

`./diff_runs BCR=123085 NDOUBLET` shows the effect on `123085`: a single cell is deleted, and
the change makes sense

`enclone TCR=74396 MOUSE`: no change; likewise for `enclone TCR=76829 MOUSE`

`enclone BCR=70838 MOUSE`: a bunch of three and four chain clonotypes are killed; no effect
on two chain clonotypes

`enclone BCR=77990 MOUSE`: minuscule effect

### CODE REVIEW

**Review by wyattmcdonnell — APPROVED** [review]
Good to go!

**Comment by DavidBJaffe** [comment]
jenkins retest this please

File: pages/auto/heuristics.html [hunk 1]
```diff
@@ -49,6 +49,39 @@ <h1>enclone heuristics</h1>
...
+<p>The algorithm then finds triples <code>(p0, p1, p2)</code> of pure subclonotypes, for which
+the following three conditions are all satisfied:
+<ul>
+<li><code>p0</code> and <code>p1</code> share an identical CDR3 DNA sequence</li>
+<li><code>p0</code> and <code>p2</code> share an identical CDR3 DNA sequence</li>
+<li><code>p1</code> and <code>p2</code> do <i>not</i> share an identical CDR3 DNA sequence.</li>
+</ul>
+</p>
+
+<p>Finally, if <code>5 * ncells(p0) <= min( ncells(p1), ncells(p2) )</code>, the entire pure
```
Comment (sreenathkrishnan) [inline pages/auto/heuristics.html:75]: The algorithm makes sense to me, but given our expected doublet rate wonder if the factor `5` is too small?

**Comment by meryllewis** [comment]
jenkins retest this please

File: pages/auto/heuristics.html (hunk 1 above)
Comment (DavidBJaffe) [reply pages/auto/heuristics.html:75]: Possibly.  The cases which are of most concern are cases where large clonotypes are a mess.  Perhaps the ideal test would not just test the ratio....  In lieu of having a relevant example I'm
reluctant to lower the factor.

Comment (DavidBJaffe) [reply pages/auto/heuristics.html:75]: For the specific customer example, a higher ratio would have worked.

**Comment by DavidBJaffe** [comment]
jenkins retest this please

File: pages/auto/heuristics.html (hunk 1 above)
Comment (DavidBJaffe) [reply pages/auto/heuristics.html:75]: I'm going to merge the PR, but we can still chew on this and tweak the algorithm.

**Review by sreenathkrishnan — APPROVED** [review]
Sounds good!

**Pull request merged**